### DISCUSSION TITLE
OpenSSL version number when compiling

### DISCUSSION THREAD

**Discussion by kgoldman** [openssl/openssl#21619] [Q&A]
Do I misunderstand the openssl version number?

+ openssl version
OpenSSL 3.1.1 30 May 2023 (Library: OpenSSL 3.1.1 30 May 2023)

In my source, I have:

``# if OPENSSL_VERSION_NUMBER > 0x30100ff0L

``#   error Untested OpenSSL version

and it hits the error.

As I understand it, 3.1.1 becomes 30100010, which should be less than 30100ff0.

What am I doing wrong?

**Comment by paulidale** [comment]
Try printing:
- OPENSSL_VERSION_MAJOR
- OPENSSL_VERSION_MINOR
- OPENSSL_VERSION_PATCH
To make sure that the version of OpenSSL you are linking against matches what you think your linking against.

You might also want to use the `OPENSSL_VERSION_PREREQ` macro to do version checks, what you are using is marked for backward compatibility use.

**Comment by t8m** [comment]
> As I understand it, 3.1.1 becomes 30100010, which should be less than 30100ff0.

No, for 3.1.1 the version number is 0x3010001F as the 0xF at the end marks a final release. That still does not explain why the condition should be true.

**Reply by mspncp** [reply]
See also my [reply on openssl-users](https://mta.openssl.org/pipermail/openssl-users/2023-August/016569.html):

> #if OPENSSL_VERSION_NUMBER > 0x30100ff0L

Kenneth, is it possible that you didn't include opensslv.h?  IIRC, the expression evaluates to #if 0 > 0x30100ff0L if OPENSSL_VERSION_NUMBER is not defined.

**Reply by kgoldman** [reply]
Thanks. 

I include evp.h, which includes macros.h, which includes opensslv.h.

What's odd is that it passes 3.0.x but not 3.1.x.  If I'm understanding the version string correctly, there must be a different issue in my build.
